### PR TITLE
Script (.py) and Notebook(.ipynb) for qubit rotation example

### PR DESCRIPTION
**Pull request by araza6**
As per our discussion on the review of PR https://github.com/qgrad/qgrad/pull/2, I have added a .py script converted from `nbconvert`, except that I deleted the cell number comments (like `#In[2]`) in the .py script rendered. For now, I have added the notebook in the new `notebooks` folder, just so that we have both the notebook as well as the script.

### CODE REVIEW

File: examples/QubitRotation.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
+#!/usr/bin/env python
+# coding: utf-8
+
+# # Qubit Rotation with JAX
+# 
+# In this notebook, we shall investigate how we can use simple gradient descent to optimize for the a qubit rotation. The idea is as follows: We start in $|1 \rangle$ state on a Bloch Sphere and apply rotations around Z-Y-Z axes (in that order), each of the three with angles $\phi$, $\theta$, and $\omega$ respectively
+#
```
Comment (quantshah) [inline examples/QubitRotation.py:7]: Add a simple line to describe what this R operator is. Eg.

"A general unitary operator is parameterised by three real parameters as, "

File: examples/QubitRotation.py [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+# 
+# $R(\phi,\theta,\omega) = RZ(\omega)RY(\theta)RZ(\phi)= \begin{bmatrix}
+# e^{-i(\phi+\omega)/2}\cos(\theta/2) & -e^{i(\phi-\omega)/2}\sin(\theta/2) \\
+# e^{-i(\phi-\omega)/2}\sin(\theta/2) & e^{i(\phi+\omega)/2}\cos(\theta/2)
+# \end{bmatrix}$
+# 
+# to reach the target state, which is defined to be $|0 \rangle$ in this case.
+
+
+import jax.numpy as jnp
```
Comment (quantshah) [inline examples/QubitRotation.py:16]: It is good practice to keep the imports from one library together leave a gap and import the other things in another block. Also import * is not recommended. Import specifically what you use in the code so that the user can clearly see where a particular function comes from. If you do from `qutip import *` and also from `qgrad import *` and use a function `fidelity()` then we do not know if it came from qutip or qgrad.

```
import jax.numpy as jnp
from jax import grad, jit
from jax.random import normal, PRNGKey


import numpy as onp


from qutip import fidelity, Qobj
```

File: examples/QubitRotation.py [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+import numpy as onp
+from qutip import *
+from jax import grad, jit
+from jax.random import normal, PRNGKey
+
+
+def rot(phi, theta, omega):
```
Comment (quantshah) [inline examples/QubitRotation.py:23]: Very well written doctoring. Great!

File: examples/QubitRotation.py [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+        rm: array[complex]
+             array representing the rotation matrix 
+    """
+    
+    cos = jnp.cos(theta/2)
+    sin = jnp.sin(theta/2)
+
+    return jnp.array(
+    [
+        [jnp.exp(-0.5j*(phi+omega))*cos, -(jnp.exp(0.5j*(phi-omega)))*sin],
```
Comment (quantshah) [inline examples/QubitRotation.py:48]: pep8 recommends whitespaces around operators such as + or - , see https://www.python.org/dev/peps/pep-0008/#other-recommendations

Since this is a fresh repo consider using Black or some other pep8 linter: https://github.com/psf/black

File: examples/QubitRotation.py [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+             
+        ket: array[complex] 
+             array representing the ket to be acted upon by by the rotation matrix
+
+    Returns:
+    -------
+        float: 
+        fidelity between the :math:`|0 \rangle` state and the evolved state under rotation    
+    """
+
```
Comment (quantshah) [inline examples/QubitRotation.py:77]: Unnecessary gap.

File: examples/QubitRotation.py [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+for epoch in range (epochs):
+    iters = 0
+    diff = 1
+    tol = 1e-7
+    while jnp.all(diff > tol) and iters < max_iters:
+        prev_weights = weights
+        der = jnp.asarray(der_cost(*prev_weights.T,init_ket))
+        weights = weights + alpha*der
+        iters += 1
+        diff = jnp.absolute(weights-prev_weights)
```
Comment (quantshah) [inline examples/QubitRotation.py:109]: This block seems to do nothing. If you see the iters, the while loop exists always after each iteration. I understand the convergence criteria that you want to use is that if the weights do not change too much then break out of the loop. However I do not see the necessity of making two blocks here using epochs and the inner loop. You could just use one of the loops and it still works.

File: examples/QubitRotation.py [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+             array representing the ket to be acted upon by by the rotation matrix
+
+    Returns:
+    -------
+        float: 
+        fidelity between the :math:`|0 \rangle` state and the evolved state under rotation    
+    """
+    
+    evolved = jnp.dot(rot(phi, theta, omega), ket)
+    return jnp.absolute(jnp.vdot(evolved.T,basis(2,0).full()))
```
Comment (quantshah) [inline examples/QubitRotation.py:79]: You must use spaces after "," see pep8 recommendations.

File: examples/QubitRotation.py [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+epochs = 100
+alpha = 0.1 #learning rate
+tol = 1e-7
+diff = 1 # diff of new and prev weights should be less than diff
+max_iters = 1
+iters = 0
+key = PRNGKey(0) #Generate a PRNG key
+weights = normal(key=key, shape=(3,))
+init_ket = basis(2,1).full()
+der_cost = grad(cost, argnums = [0,1,2])
```
Comment (quantshah) [inline examples/QubitRotation.py:98]: You must use spaces after "," see pep8 recommendations.

File: examples/QubitRotation.py [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+
+for epoch in range (epochs):
+    iters = 0
+    diff = 1
+    tol = 1e-7
+    while jnp.all(diff > tol) and iters < max_iters:
+        prev_weights = weights
+        der = jnp.asarray(der_cost(*prev_weights.T,init_ket))
```
Comment (quantshah) [inline examples/QubitRotation.py:106]: You must use spaces after "," see pep8 recommendations.

File: examples/QubitRotation.py [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+
+
+# ## Bloch Sphere Visualization
+# 
+# We see that started off with a very low fidelity. With gradient descent iterations, we progressively achieve better fidelities via better parameters,$\phi$, $\theta$, and $\omega$. To see it visually, we render our states on to a Bloch sphere.
+# 
+# On a Bloch sphere below, we started with $|1 \rangle$ state (green arrow) and the optimized parameters of the rotation matrix give us our desired $R(\phi,\theta,\omega)$ such that $R(\phi,\theta,\omega) |1 \rangle$ evolves the state $|1 \rangle$ to the state represented by the brown arrow, which is very close to our target state, $|0 \rangle$ (blue arrow), as desired.
+
+
+b = Bloch()
```
Comment (quantshah) [inline examples/QubitRotation.py:123]: Maybe plot the fidelities as training progresses? That is easier to see the progress instead of printing fidelities. Good to see the bloch sphere representation.

File: examples/QubitRotation.py (hunk 6 above)
Comment (araza6) [reply examples/QubitRotation.py:109]: I made the epochs loop here just in case a user wants to use epochs. Better code reuse

**Comment by araza6** [comment]
If @quantshah gives a pass, I'll merge this

**Comment by araza6** [comment]
@nathanshammah and @quantshah, just a ping. This PR is review-ready as well

File: examples/QubitRotation.py (hunk 7 above)
Comment (nathanshammah) [reply examples/QubitRotation.py]: @araza6 please mark the conversations as resolved (if you see the option below, so that reviewers are sure this is taken care of) or comment each of these comments below.

**Comment by nathanshammah** [comment]
I am merging this PR. @araza6 can you please open two issues about my two comments above (color gradient and checkpoint) so that we can keep track of this? Thanks.
